flink-tests/src/test/java/org/apache/flink/test/recovery/BatchFineGrainedRecoveryITCase.java

```diff
@@ -101,101 +166,276 @@ public void teardown() throws Exception {
 
-		private final CoinToss coin;
+		private ExceptionFailureStrategy(int failAfterCallNumber) {
+			super(failAfterCallNumber);
+		}
 
-		private RandomExceptionFailureStrategy(int probFraction, int probBase) {
-			this.coin = new CoinToss(probFraction, probBase);
+		@Override
+		void fail(int trackingIndex) throws FlinkException {
+			throw new FlinkException("BAGA-BOOM!!! The user function generated test failure.");
+		}
+	}
+
+	private static class TaskExecutorFailureStrategy extends AbstractOnceAfterCallNumberFailureStrategy {
+		private static final long serialVersionUID = 1L;
+
+		private TaskExecutorFailureStrategy(int failAfterCallNumber) {
+			super(failAfterCallNumber);
 		}
 
 		@Override
-		public void failOrNot() {
-			if (coin.toss() && StaticFailureCounter.failOrNot()) {
-				throw new FlinkRuntimeException("BAGA-BOOM!!! The user function generated test failure.");
+		void fail(int trackingIndex) throws Exception {
+			//noinspection OverlyBroadCatchBlock
+			try {
+				restartTaskManager();
+			} catch (InterruptedException e) {
+				// ignore the exception, task should have been failed while stopping TM
+				Thread.currentThread().interrupt();
+			} catch (Throwable t) {
+				StaticMapFailureTracker.unrelatedFailure(t);
+				throw t;
 			}
 		}
 	}
 
-	private static class CoinToss implements Serializable {
+	private abstract static class AbstractOnceAfterCallNumberFailureStrategy implements FailureStrategy {
 		private static final long serialVersionUID = 1L;
-		private static final Random rnd = new Random();
 
-		private final int probFraction;
-		private final int probBase;
+		private final UUID id;
+		private final int failAfterCallNumber;
+		private transient int callCounter;
+
+		private AbstractOnceAfterCallNumberFailureStrategy(int failAfterCallNumber) {
+			this.failAfterCallNumber = failAfterCallNumber;
+			id = UUID.randomUUID();
+		}
+
+		@Override
+		public boolean failOrNot(int trackingIndex) throws Exception {
+			callCounter++;
+			boolean generateFailure = callCounter == failAfterCallNumber;
+			if (generateFailure) {
+				fail(trackingIndex);
+			}
+			return generateFailure;
+		}
+
+		abstract void fail(int trackingIndex) throws Exception;
+
+		@Override
+		public String toString() {
+			return this.getClass().getSimpleName() + " (fail after " + failAfterCallNumber + " calls)";
+		}
+
+		@Override
+		public boolean equals(Object o) {
+			if (this == o) {
+				return true;
+			}
+			if (o == null || getClass() != o.getClass()) {
+				return false;
+			}
+			return Objects.equals(id, ((AbstractOnceAfterCallNumberFailureStrategy) o).id);
+		}
+
+		@Override
+		public int hashCode() {
+			return id.hashCode();
+		}
+	}
+
+	@SuppressWarnings("SynchronizationOnStaticField")
+	private enum StaticMapFailureTracker {
+		;
+
+		private static final List<AtomicInteger> mapRestarts = new ArrayList<>(MAP_NUMBER);
+		private static final List<Map<FailureStrategy, Boolean>> mapFailures = new ArrayList<>(MAP_NUMBER);
```

Comment:
instead of a map you could use a set, and only add the strategy if it has failed.